src/vos/vos_query.c

```diff
@@ -322,28 +335,51 @@ vos_obj_query_key(daos_handle_t coh, daos_unit_oid_t oid, uint32_t flags,
 		return -DER_INVAL;
 	}
 
+	query.qt_ts_set = NULL;
+
 	if (flags & VOS_GET_DKEY) {
 		if (dkey == NULL) {
 			D_ERROR("dkey can't be NULL with VOS_GET_DKEY\n");
 			return -DER_INVAL;
 		}
 		daos_anchor_set_zero(&dkey_anchor);
+
+		cflags = VOS_TS_READ_OBJ;
 	}
 
 	if (flags & VOS_GET_AKEY && akey == NULL) {
 		D_ERROR("akey can't be NULL with VOS_GET_AKEY\n");
 		return -DER_INVAL;
+
+		if (cflags == 0)
+			cflags = VOS_TS_READ_DKEY;
 	}
 
 	if (flags & VOS_GET_RECX && recx == NULL) {
 		D_ERROR("recx can't be NULL with VOS_GET_RECX\n");
 		return -DER_INVAL;
+
+		nr_akeys = 1;
+		if (cflags == 0)
+			cflags = VOS_TS_READ_AKEY;
 	}
 
 	vos_dth_set(dth);
+	rc = vos_ts_set_allocate(&query.qt_ts_set, 0, cflags, nr_akeys,
```

Comment:
Where is related ts_set_free() is called?

Reply:
good catch :-)